src/DDM.jl

```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+#         end
+#     elseif rt - τ + st/2 <= min_rt # is t lower than lower boundary Ter distr?
+#         Fnew = 0
+#     end
+    
+#     Fnew = Fnew > Δ ? Fnew : 0
+
+#     return    Fnew
+
+# end
+
 """
     rand(dist::DDM)
 
-Generate a random rt for the Diffusion Decision Model (negative coding)
+Generate a random choice and rt for the Diffusion Decision Model
 
 # Arguments
-- `dist`: model object for the Diffusion Decision Model. 
+- `dist`: model object for Diffusion Decision Model. 
+    
+method simulating the diffusion process: 
+_rand_rejection uses Tuerlinckx et al., 2001 rejection-based method for the general wiener process
+_rand_stochastic uses the stochastic Euler method to simulate the stochastic differential equation
+
+# References
+
+    Tuerlinckx, F., Maris, E., Ratcliff, R., & De Boeck, P. (2001). 
+    A comparison of four methods for simulating the diffusion process. 
+    Behavior Research Methods, Instruments, & Computers, 33, 443-456.
+
+    Converted from Rhddmjagsutils.R R script by Kianté Fernandez
+    See also https://github.com/kiante-fernandez/Rhddmjags.
 """
 function rand(rng::AbstractRNG, d::DDM)
     return _rand_rejection(rng, d)
+    # return _rand_stochastic(rng, d)
 end
 
-# Rejection-based Method for the Symmetric Wiener Process(Tuerlinckx et al., 2001 based on Lichters et al., 1995)
-# adapted from the RWiener R package, note, here σ = 0.1
-function _rand_rejection(rng::AbstractRNG, d::DDM)
-    (ν, α, τ, z) = params(d)
-    
-    ϵ = 1.0e-15
-
-    D = .005 # = 2*σ^2 => 1/200
-    zn = (z*α) / 10 # absolute bias!
-    αn = α / 10
-    νn = ν / 10
-
-    total_time = 0.0
-    start_pos = 0.0
-    Aupper = αn - zn
-    Alower = -zn
-    radius = min(abs(Aupper), abs(Alower))
-    λ = 0.0
-    F = 0.0
-    prob = 0.0
-
-    while true
-        if νn == 0
-            λ = (0.25D*π^2) / (radius^2)
-            F = 1.0
-            prob = .5            
-        else
-            λ = ((0.25*νn^2)/D) + ((0.25*D*π^2) / (radius^2))
-            F = (D*π) / (radius*νn)
-            F = F^2 / (1+F^2)
-            prob = exp((radius*νn)/D)
-            prob = prob / (1+prob)
-        end
+function _rand_rejection(rng::AbstractRNG, d::DDM; N::Int = 1)
+    (ν, α, τ, z, η, sz, st, σ) = params(d)
 
-        r = rand(rng)
-        dir = r < prob ? 1 : -1
-        l = -1.0
-        s1 = 0.0
-        s2 = 0.0
-
-        # Tuerlinckx et al. (2001; eq. 16)  
-        while s2 > l
-            s1 = rand(rng)
-            s2 = rand(rng)
-            tnew = 0.0
-            tδ = 0.0
-            uu = zero(Int)
-
-            while (abs(tδ) > ϵ) || (uu == 0)
-                uu += 1
-                tt = 2*uu + 1
-                tδ = tt * (uu % 2 == 0 ? 1 : -1) * (s1 ^ (F * tt^2))
-                tnew += tδ                              
-            end
+    if η == 0
+        η = 1e-16
+    end
 
-            l = 1 + (s1^(-F)) * tnew 
+    # Initialize output vectors
```

Comment:
I made a few stylistic changes, and cleaned up some redundant variables. Can you use the follow if it looks OK to you? 


```
function _rand_rejection(rng::AbstractRNG, d::RatcliffDDM; N::Int = 1)
    (;ν, α, τ, z, η, sz, st, σ) = d

    η = η == 0 ? eps() : η

    # Initialize output vectors
    result = zeros(N)
    T = zeros(N)
    XX = fill(0, N)

    # Called sigma in 2001 paper
    D = σ^2 / 2

    # Program specifications
    ϵ = eps()  # precision from 1.0 to next double-precision number

    for n in 1:N
        r1 = randn()
        μ = ν + r1 * η
        bb = z - sz / 2 + sz * rand()
        zz = bb * α
        finished = false
        totaltime = 0.0
        startpos = 0.0
        Aupper = α - zz
        Alower = -zz
        radius = min(abs(Aupper), abs(Alower))

        while !finished
            λ = 0.25 * μ^2 / D + 0.25 * D * π^2 / radius^2
            # eq. formula (13) in 2001 paper with D = sigma^2/2 and radius = Alpha/2
            F = D * π / (radius * μ)
            F = F^2 / (1 + F^2)
            # formula p447 in 2001 paper
            prob = exp(radius * μ / D)
            prob = prob / (1 + prob)
            dir_ = 2 * (rand() < prob) - 1
            l = -1.0
            s2 = 0.0
            s1 = 0.0
            while s2 > l
                s2 = rand()
                s1 = rand()
                tnew = 0.0
                told = 0.0
                uu = 0
                while abs(tnew - told) > ϵ || uu == 0
                    told = tnew
                    uu += 1
                    tnew = told + (2 * uu + 1) * (-1)^uu * s1^(F * (2 * uu + 1)^2)
                    # infinite sum in formula (16) in BRMIC,2001
                end
                l = 1 + s1^(-F) * tnew
            end
            # rest of formula (16)
            t = abs(log(s1)) / λ
            # is the negative of t* in (14) in BRMIC,2001
            totaltime += t
            dir_ = startpos + dir_ * radius
            ndt = τ - st / 2 + st * rand()
            if (dir_ + ϵ) > Aupper
                T[n] = ndt + totaltime
                XX[n] = 1
                finish = true
            elseif (dir_ - ϵ) < Alower
                T[n] = ndt + totaltime
                XX[n] = 2
                finish = true
            else
                startpos = dir_
                radius = minimum(abs.([Aupper, Alower] .- startpos))
            end
        end
    end
    return (choice=XX,rt=T)
end
```

Reply:
Looks fine to me. I made the corresponding changes.